US618160: Ingestion Workflow: Ignore unexpected fields sent by the agent

jonny-mccoubrey opened the pull request:
https://internal.almoctane.com/ui/entity-navigation?p=131002/6001&entityType=work_item&id=618160

jonny-mccoubrey committed:
Implement FieldValidator

jonny-mccoubrey committed:
Formatting

buildmachine-sou-jenkins2 commented:
A developer build has not yet been created for this branch. Click [here](http://sou-jenkins2.swinfra.net:8080/job/Create-New-Build/parambuild/?GITHUB_SERVER=github.com&GITHUB_ORGANISATION=CAFDataProcessing&REPO_NAME=staging-service&BRANCH_NAME=US618160) to go ahead and create the build...

buildmachine-sou-jenkins2 commented:
CI Build Link:
http://sou-jenkins2.swinfra.net:8080/job/CAFDataProcessing/job/CAFDataProcessing~staging-service~US618160~CI

jonny-mccoubrey committed:
Add logic for flattened fields in adapter

jonny-mccoubrey committed:
Formatting

jonny-mccoubrey committed:
Validate against flattened fields in document

jonny-mccoubrey committed:
Testing

jonny-mccoubrey committed:
Refactoring

jonny-mccoubrey committed:
Wire up validator to IngestionBatchWorkerPlugin

jonny-mccoubrey committed:
Set var to final

jonny-mccoubrey committed:
Renamed Adapter and AgentFieldAdapter

jonny-mccoubrey committed:
Re-structured validator code

jonny-mccoubrey committed:
Final var

jonny-mccoubrey committed:
Instantiate validator in IngestionBatchWorkerPlugin constructor

dermot-hardy committed:
Merge branch 'develop' into US618160

dermot-hardy commented on pom.xml:
```diff
@@ -333,6 +344,11 @@
+                <groupId>org.json</groupId>
+                <artifactId>json</artifactId>
+                <version>20220320</version>
+            </dependency>
```
- Dependency management should be in alphabetical order, so if we're adding this then it should be up between `org.apache.tomcat.embed:tomcat-embed-core` and `org.mockito:mockito-core`.
- More importantly, I'm not particularly comfortable with the introduction of this dependency, both because it has a non-standard license and we want to get rid of it from other projects (see [`US354614`](https://internal.almoctane.com/ui/entity-navigation?p=131002/6001&entityType=work_item&id=354614)), and because we already use [Jackson](https://github.com/FasterXML/jackson) for handling JSON in the worker so I'm not sure why we would introduce a different JSON processing library.

dermot-hardy commented on pom.xml:
```diff
@@ -157,6 +157,17 @@
             </dependency>
+            <dependency>
+                <groupId>com.github.stefanbirkner</groupId>
+                <artifactId>system-rules</artifactId>
```
Dependency management should be in alphabetical order.

dermot-hardy commented on pom.xml:
```diff
@@ -157,6 +157,17 @@
+                        <groupId>org.hamcrest</groupId>
+                        <artifactId>hamcrest-core</artifactId>
+                    </exclusion>
+                </exclusions>
```
Why do we need this exclusion?

jonny-mccoubrey replied:
![dependencyIssue](https://user-images.githubusercontent.com/44871582/225291908-2e60e451-b3b7-4f99-b4a2-c16aecb06746.JPG)


Getting this dependency problem when it is not excluded

dermot-hardy commented on worker-batch-ingestion-plugin/pom.xml:
```diff
@@ -68,6 +68,11 @@
+            <groupId>com.github.stefanbirkner</groupId>
+            <artifactId>system-rules</artifactId>
+            <scope>test</scope>
+        </dependency>
```
New test-scope dependencies should be down with the other test-scope dependencies, not up here with the compile-scope dependencies.

dermot-hardy commented on worker-batch-ingestion-plugin/pom.xml:
```diff
@@ -131,6 +136,10 @@
+        <dependency>
+            <groupId>org.json</groupId>
+            <artifactId>json</artifactId>
+        </dependency>
```
- New compile-scope dependencies should be up with the other compile-scope dependencies, not down here with the test-scope dependencies.
- I already mentioned my reservations about the introduction of this dependency in an earlier comment.

dermot-hardy commented on worker-batch-ingestion-plugin/src/main/java/com/github/cafdataprocessing/worker/ingestion/IngestionBatchWorkerPlugin.java:
```diff
@@ -88,6 +92,20 @@ public IngestionBatchWorkerPlugin()
                                      new DocumentWorkerDocumentDeserializer(totalSubdocumentLimit.get()));
         mapper.registerModule(simpleModule);
+
+        validationFile = System.getenv("CAF_VALIDATION_FILE");
```
Let's make this `CAF_INGESTION_BATCH_WORKER_VALIDATION_FILE`.

dermot-hardy commented on worker-batch-ingestion-plugin/src/main/java/com/github/cafdataprocessing/worker/ingestion/validator/FieldValidator.java:
```diff
@@ -0,0 +1,84 @@
+public class FieldValidator
+{
+    final Set<String> allowedFields;
+    final Map<String, Set<String>> allowedFlattenedFields;
```
Instance variables should usually be `private`.

```suggestion
    private final Set<String> allowedFields;
    private final Map<String, Set<String>> allowedFlattenedFields;
```

dermot-hardy commented on worker-batch-ingestion-plugin/src/main/java/com/github/cafdataprocessing/worker/ingestion/IngestionBatchWorkerPlugin.java:
```diff
@@ -64,6 +66,8 @@ public final class IngestionBatchWorkerPlugin implements BatchWorkerPlugin
     private final ObjectMapper mapper;
     private final BatchPathProvider fileSystemProvider;
+    private final String validationFile;
+    private final FieldValidator fieldValidator;
```
Rather than storing both of these and creating a dummy unused field validator when the file is not specified, create an interface for your field validator to implement, and then create a singleton implementation of it that does nothing.  You can then set the field validator to `NullFieldValidator.INSTANCE` if the validation file is not specified.

    public interface FieldValidatorInterface
    {
        DocumentWorkerDocument validate(DocumentWorkerDocument document);
    }

    public enum NullFieldValidator implements FieldValidatorInterface
    {
        INSTANCE;

        @Override
        public DocumentWorkerDocument validate(final DocumentWorkerDocument document)
        {
            return document;
        }

        @Override
        public String toString()
        {
            return "<null field validator>";
        }
    }

```suggestion
    private final FieldValidatorInterface fieldValidator;
```

dermot-hardy commented on worker-batch-ingestion-plugin/src/main/java/com/github/cafdataprocessing/worker/ingestion/validator/ValidationFileAdapter.java:
```diff
@@ -0,0 +1,79 @@
+import java.util.Map;
+import java.util.Set;
+
+public class ValidationFileAdapter
```
Classes shouldn't be `public` unnecessarily but they should be `final`.

```suggestion
final class ValidationFileAdapter
```

dermot-hardy commented on worker-batch-ingestion-plugin/src/main/java/com/github/cafdataprocessing/worker/ingestion/validator/ValidationFileAdapterException.java:
```diff
@@ -0,0 +1,24 @@
+ */
+package com.github.cafdataprocessing.worker.ingestion.validator;
+
+public class ValidationFileAdapterException extends Exception
```
Classes should usually be `final`.

```suggestion
public final class ValidationFileAdapterException extends Exception
```

dermot-hardy replied:
So far as I can see, this exception is only thrown if there is an I/O error reading the file.  Its name is overly generic for that purpose.  I'd suggest making it much more specific by renaming it to `FailedToReadValidationFileException`.  Alternatively you could get rid of it altogether and allow the `IOException` to bubble up to `IngestionBatchWorkerPlugin` in its place, given that it's not really adding any value.

dermot-hardy commented on worker-batch-ingestion-plugin/src/main/java/com/github/cafdataprocessing/worker/ingestion/validator/ValidationFileAdapterException.java:
```diff
@@ -0,0 +1,24 @@
+
+public class ValidationFileAdapterException extends Exception
+{
+    ValidationFileAdapterException(final String message, final Throwable cause)
```
Good call not making this `public`.

dermot-hardy commented on worker-batch-ingestion-plugin/src/main/java/com/github/cafdataprocessing/worker/ingestion/IngestionBatchWorkerPlugin.java:
```diff
@@ -88,6 +92,20 @@ public IngestionBatchWorkerPlugin()
+                fieldValidator = new FieldValidator(validationFile);
+            } catch (final ValidationFileAdapterException validationFileAdapterException) {
+                log.error("Exception when attempting to read validation file" + "\n"
+                    + validationFileAdapterException.getMessage());
```
Don't customize exception logging.  Use the overload that takes a message and the exception, so that the logging library can handle the formatting of the exception properly.

```suggestion
                log.error("Exception when attempting to read validation file", validationFileAdapterException);
```

dermot-hardy commented on worker-batch-ingestion-plugin/src/main/java/com/github/cafdataprocessing/worker/ingestion/IngestionBatchWorkerPlugin.java:
```diff
@@ -88,6 +92,20 @@ public IngestionBatchWorkerPlugin()
+        if (StringUtils.isNotEmpty(validationFile)) {
+            try {
+                fieldValidator = new FieldValidator(validationFile);
+            } catch (final ValidationFileAdapterException validationFileAdapterException) {
```
Usually you should just use `ex` for exception variables.  They normally have a very small scope.

```suggestion
            } catch (final ValidationFileAdapterException ex) {
```

dermot-hardy commented on worker-batch-ingestion-plugin/src/main/java/com/github/cafdataprocessing/worker/ingestion/IngestionBatchWorkerPlugin.java:
```diff
@@ -88,6 +92,20 @@ public IngestionBatchWorkerPlugin()
+                log.error("Exception when attempting to read validation file" + "\n"
+                    + validationFileAdapterException.getMessage());
+                throw new RuntimeException("Exception when attempting to read validation file" + "\n"
+                    + validationFileAdapterException.getMessage());
```
Generally I would avoid putting newlines in exception messages.  Also, when re-throwing an exception, pass the original exception up as the cause:
```suggestion
                throw new RuntimeException("Exception when attempting to read validation file: "
                    + validationFileAdapterException.getMessage(), validationFileAdapterException);
```

dermot-hardy commented on worker-batch-ingestion-plugin/src/main/java/com/github/cafdataprocessing/worker/ingestion/IngestionBatchWorkerPlugin.java:
```diff
@@ -227,6 +245,12 @@ private DocumentWorkerDocumentTask createDocument(final String line, final Map<S
         throws IOException, BatchDefinitionException
     {
         final DocumentWorkerDocumentTask document = mapper.readValue(line, DocumentWorkerDocumentTask.class);
+        document.document.failures = new ArrayList<>();
```
This line would wipe out any failures that are already on the document.  That is not invalid.  If the agent has an issue (e.g. the text could not be extracted because the file format was unknown) then the document staged will already have some failures.

dermot-hardy commented on worker-batch-ingestion-plugin/src/main/java/com/github/cafdataprocessing/worker/ingestion/validator/FieldValidator.java:
```diff
@@ -0,0 +1,84 @@
+import java.util.Map;
+import java.util.Set;
+
+public class FieldValidator
```
Classes should usually be `final`.

```suggestion
public final class FieldValidator
```

dermot-hardy commented on worker-batch-ingestion-plugin/src/main/java/com/github/cafdataprocessing/worker/ingestion/validator/FieldValidator.java:
```diff
@@ -0,0 +1,84 @@
+
+    public FieldValidator(final String validationFile) throws ValidationFileAdapterException
+    {
+        ValidationFileAdapter adapter = new ValidationFileAdapter(validationFile);
```
Missing `final` in variable declaration.

```suggestion
        final ValidationFileAdapter adapter = new ValidationFileAdapter(validationFile);
```

dermot-hardy commented on worker-batch-ingestion-plugin/src/main/java/com/github/cafdataprocessing/worker/ingestion/validator/ValidationFileAdapter.java:
```diff
@@ -0,0 +1,79 @@
+    public Map<String, Set<String>> getFlattenedFieldKeys()
+    {
+        final Map<String, Set<String>> flattenedFields = new HashMap<>();
+        for (String fieldKey : getFieldKeys()) {
```
Loop variables should also be `final`:

```suggestion
        for (final String fieldKey : getFieldKeys()) {
```

dermot-hardy commented on worker-batch-ingestion-plugin/src/main/java/com/github/cafdataprocessing/worker/ingestion/validator/FieldValidator.java:
```diff
@@ -0,0 +1,84 @@
+        allowedFlattenedFields = adapter.getFlattenedFieldKeys();
+    }
+
+    public DocumentWorkerDocument validate(final DocumentWorkerDocument document)
```
This method is only looking at the document's fields, but the document could also have subdocuments (i.e. if the document is an e-mail then any attachments that it has will be subdocuments).  There needs to be recursive validation of the subdocuments.

dermot-hardy commented on worker-batch-ingestion-plugin/src/main/java/com/github/cafdataprocessing/worker/ingestion/validator/FieldValidator.java:
```diff
@@ -0,0 +1,84 @@
+        for (final String key : keySet) {
+            if (!(isValidField(key) || isValidFlattenedField(key))) {
+                final DocumentWorkerFailure fieldNotAllowedFailure = new DocumentWorkerFailure();
+                fieldNotAllowedFailure.failureId = "FIELD-NOT-ALLOWED-FAILURE";
```
Indexing Worker error codes: [ElasticWorkerError.java](https://github.houston.softwaregrp.net/caf/worker-elastic/blob/v3.0.0/worker-elastic-core/src/main/java/com/microfocus/caf/worker/elastic/core/ElasticWorkerError.java#L27) &ndash; `EW-0001`, `EW-0002`, etc.
AppResources Worker error codes: [AppResourcesWorkerError.java](https://github.houston.softwaregrp.net/Verity/darwin-worker-appresources/blob/v4.0.0/worker-appresources/src/main/java/com/microfocus/apollo/workers/appresources/AppResourcesWorkerError.java#L26) &ndash; `APW-0001`, `APW-0002`, `APW-C101`, etc.

Just follow a similar pattern, so `IW-0001` or something along those lines.

dermot-hardy commented on worker-batch-ingestion-plugin/src/main/java/com/github/cafdataprocessing/worker/ingestion/validator/FieldValidator.java:
```diff
@@ -0,0 +1,84 @@
+            if (!(isValidField(key) || isValidFlattenedField(key))) {
+                final DocumentWorkerFailure fieldNotAllowedFailure = new DocumentWorkerFailure();
+                fieldNotAllowedFailure.failureId = "FIELD-NOT-ALLOWED-FAILURE";
+                fieldNotAllowedFailure.failureMessage
+                    = String.format(key + " is not allowed to be set by the agent");
```
You're clearly didn't mean to use the `String.format()` here, and you don't need it either.

```suggestion
                fieldNotAllowedFailure.failureMessage = key + " is not allowed to be set by the agent";
```

dermot-hardy commented on worker-batch-ingestion-plugin/src/main/java/com/github/cafdataprocessing/worker/ingestion/validator/FieldValidator.java:
```diff
@@ -0,0 +1,84 @@
+                fieldNotAllowedFailure.failureId = "FIELD-NOT-ALLOWED-FAILURE";
+                fieldNotAllowedFailure.failureMessage
+                    = String.format(key + " is not allowed to be set by the agent");
+                document.failures.add(fieldNotAllowedFailure);
```
You should be prepared to deal with an [`UnsupportedOperationException`](https://docs.oracle.com/javase/8/docs/api/java/lang/UnsupportedOperationException.html) if the list of failures is not mutable.

dermot-hardy commented on worker-batch-ingestion-plugin/src/main/java/com/github/cafdataprocessing/worker/ingestion/validator/FieldValidator.java:
```diff
@@ -0,0 +1,84 @@
+                fieldNotAllowedFailure.failureMessage
+                    = String.format(key + " is not allowed to be set by the agent");
+                document.failures.add(fieldNotAllowedFailure);
+                document.fields.remove(key);
```
You should be prepared to deal with an [`UnsupportedOperationException`](https://docs.oracle.com/javase/8/docs/api/java/lang/UnsupportedOperationException.html) if the map of fields is not mutable.

dermot-hardy commented on worker-batch-ingestion-plugin/src/main/java/com/github/cafdataprocessing/worker/ingestion/validator/FieldValidator.java:
```diff
@@ -0,0 +1,84 @@
+
+    private boolean isValidField(final String fieldKey)
+    {
+        for (final String allowedField : allowedFields) {
+            if (fieldKey.equals(allowedField)) {
+                return true;
+            }
+        }
+        return false;
```
This is very inefficient (because if the way the method is called it is O(n&sup2;) in Big O notation).  You should be using the `.contains()` method instead:

```suggestion
        return allowedFields.contains(fieldKey);
```

dermot-hardy commented on worker-batch-ingestion-plugin/src/main/java/com/github/cafdataprocessing/worker/ingestion/validator/ValidationFileAdapter.java:
```diff
@@ -0,0 +1,79 @@
+        return flattenedFields;
+    }
+
+    static String getFileContents(final String filePath) throws ValidationFileAdapterException
```
This method is not marked `private`.

```suggestion
    private static String getFileContents(final String filePath) throws ValidationFileAdapterException
```

dermot-hardy replied:
The JSON library should have methods that allow you to read the JSON directly from a file, without first having to go through a String.  It probably doesn't matter too much in this case but generally that will be the most memory-efficient approach.

dermot-hardy commented on worker-batch-ingestion-plugin/src/main/java/com/github/cafdataprocessing/worker/ingestion/validator/ValidationFileAdapter.java:
```diff
@@ -0,0 +1,79 @@
+
+    public Set<String> getFieldKeys()
+    {
+        return fieldsJsonObject.keySet();
```
Actually the `objectEncoding=flattened` type fields would not be valid standalone.

jonny-mccoubrey replied:
Do you mean that [this](https://github.com/CAFDataProcessing/staging-service/blob/c772c80a80f7d6a14e19e592fcec0983a1c3305d/worker-batch-ingestion-plugin/src/main/java/com/github/cafdataprocessing/worker/ingestion/validator/ValidationFileAdapter.java#L64) is not valid?

dermot-hardy replied:
As discussed, no, I meant that `OCR` for example would not be valid by itself, because it is a flattened field.

dermot-hardy commented on worker-batch-ingestion-plugin/src/main/java/com/github/cafdataprocessing/worker/ingestion/validator/ValidationFileAdapter.java:
```diff
@@ -0,0 +1,79 @@
+        for (String fieldKey : getFieldKeys()) {
+            final JSONObject field = fieldsJsonObject.getJSONObject(fieldKey);
+            if (field.optString("objectEncoding").equals("flattened")) {
+                final JSONObject propertiesJsonObject = typesJsonObject.getJSONObject(fieldKey.toLowerCase());
```
It's not the field key you need to look at; it is the type.  Also never use `.toLowerCase()` without specifying a character set; you would get different behaviour on different platforms.

jonny-mccoubrey replied:
Have looked and not sure what character set needs to be specified for `.toLowerCase()`.
Have also now introduced `.toUpperCase()` which will probably need the same thing.

dermot-hardy commented on worker-batch-ingestion-plugin/src/main/java/com/github/cafdataprocessing/worker/ingestion/validator/ValidationFileAdapter.java:
```diff
@@ -0,0 +1,79 @@
+        return fieldsJsonObject.keySet();
+    }
+
+    public Map<String, Set<String>> getFlattenedFieldKeys()
```
I'm not convinced that this output structure actually covers the recursive possibilities here - though I'm slightly torn because I also don't want the code to be too complicated.

jonny-mccoubrey replied:
Have refactored logic to regex - see comment below.

jonny-mccoubrey committed:
Address PR review comments

jonny-mccoubrey committed:
FieldValidatorInterface and NullFieldValidator

jonny-mccoubrey committed:
Use Jackson instead of Json library

jonny-mccoubrey committed:
Address PR review comments

jonny-mccoubrey committed:
Fix getResourcesAsStream() file paths

jonny-mccoubrey committed:
Prepare for documents to have existing failures

jonny-mccoubrey committed:
Validate subDocuments

jonny-mccoubrey committed:
Prepare for UnsupportedOperationException

jonny-mccoubrey committed:
Refactoring

jonny-mccoubrey committed:
Refactor flattened field logic with regex

jonny-mccoubrey commented:
@dermot-hardy have addressed review comments and also taken a new approach to validating the flattened fields. Now generating a regex statement and checking if field keys match it.
Can discuss if this is not appropriate.

jonny-mccoubrey committed:
Use Files.newInputStream to read file

dermot-hardy committed:
Merge branch 'develop' into US618160

dermot-hardy commented on worker-batch-ingestion-plugin/src/main/java/com/github/cafdataprocessing/worker/ingestion/validator/ValidationFileAdapter.java:
```diff
@@ -0,0 +1,95 @@
+                }
+            } else {
+                newField = field.getKey();
+                fieldKeys.add(newField);
```
Mixing regular expressions with non-regular expressions is very dangerous.  You need to use [`Pattern.quote()`](https://docs.oracle.com/javase/8/docs/api/java/util/regex/Pattern.html#quote-java.lang.String-) to get a regular expression for the literal, so that only regular expressions go into the list.

jonny-mccoubrey replied:
I have added this now. This still seems to work for literal fields that I am passing in for 'non-flattened' fields. For example, the field "MIMETYPE" now looks like "\QMIMETYPE\E" in the list of allowed fields. This still seems to work - I had worried it might not.

dermot-hardy commented on worker-batch-ingestion-plugin/src/main/java/com/github/cafdataprocessing/worker/ingestion/validator/ValidationFileAdapter.java:
```diff
@@ -0,0 +1,95 @@
+            } else {
+                newField = field.getKey();
+                fieldKeys.add(newField);
+            }
```
Both `else` blocks are the same, so the `if`s can be merged to eliminate one of them.

```suggestion
            final Map.Entry<String, JsonNode> field = fieldIterator.next();
            final JsonNode objectEncodingNode = field.getValue().get("objectEncoding");
            if (objectEncodingNode != null && objectEncodingNode.asText().equals("flattened")) {
                fieldKeys.addAll(generateFlattenedFieldRegex(field));
            } else {
                final String newField = field.getKey();
                fieldKeys.add(Pattern.quote(newField));
            }
```

dermot-hardy commented on worker-batch-ingestion-plugin/src/main/java/com/github/cafdataprocessing/worker/ingestion/validator/ValidationFileAdapter.java:
```diff
@@ -0,0 +1,95 @@
+    {
+        final ArrayList<String> flattenedFields = new ArrayList<>();
+
+        final String[] fieldType = field.getValue().get("type").asText().split("(?<=\\[)|(?=\\[)", 2);
```
Is the first part of that expression required?

```suggestion
        final String[] fieldType = field.getValue().get("type").asText().split("(?=\\[)", 2);
```

dermot-hardy commented on worker-batch-ingestion-plugin/src/main/java/com/github/cafdataprocessing/worker/ingestion/validator/ValidationFileAdapter.java:
```diff
@@ -0,0 +1,95 @@
+        while (properties.hasNext()) {
+            final String suffix = properties.next().getKey();
+            final StringBuilder sb = new StringBuilder();
+            sb.append("^").append("(").append(fieldType[0].toUpperCase()).append(")");
```
`fieldType[0].toUpperCase()` is the name of the type.  It seems like a logic error to be including it in the regular expression.  I assume you meant to use `field.getKey()` here?  If so then it would still need to be wrapped with a call to `Pattern.quote()`.

dermot-hardy commented on worker-batch-ingestion-plugin/src/main/java/com/github/cafdataprocessing/worker/ingestion/validator/ValidationFileAdapter.java:
```diff
@@ -0,0 +1,95 @@
+            final StringBuilder sb = new StringBuilder();
+            sb.append("^").append("(").append(fieldType[0].toUpperCase()).append(")");
+
+            sb.append("+(_(?=\\d))");
```
- Why the `+`?  That is one **or more**.
- The keys do not have to be digits.  They can be anything that doesn't have an underscore.
- You don't seem to be allowing multiple characters.

jonny-mccoubrey replied:
New logic (example) based on these parameters:
^(\QMETADATA_FILES\E)_([^_]+)_(CONTENT)$

dermot-hardy commented on worker-batch-ingestion-plugin/src/main/java/com/github/cafdataprocessing/worker/ingestion/validator/ValidationFileAdapter.java:
```diff
@@ -0,0 +1,95 @@
+
+            sb.append("+(_(?=\\d))");
+            for (int i = 0; i < count; i++) {
+                sb.append("+((?<=_)\\d(?=_))+((?<=\\d)_)");
```
I think we need to sit down and go through this but this looks more complicated that I would expect is necessary.

jonny-mccoubrey replied:
I have simplified it by removing the positive look ahead and positive look behind which was unnecessary

dermot-hardy commented on worker-batch-ingestion-plugin/src/main/java/com/github/cafdataprocessing/worker/ingestion/validator/ValidationFileAdapter.java:
```diff
@@ -0,0 +1,95 @@
+            sb.append("+").append("(").append(suffix.toUpperCase()).append(")").append("$");
+
+            flattenedFields.add(sb.toString());
+            property.fields().next();
```
There is no recursion here.  We can probably get away with that for now because the only recursive field we have (i.e. `COLUMNS`.`GRAMMAR_MATCHES`) is `json` rather than `flattened`, but in principle this code should be recursive.

dermot-hardy commented on worker-batch-ingestion-plugin/src/main/java/com/github/cafdataprocessing/worker/ingestion/validator/FieldValidator.java:
```diff
@@ -0,0 +1,79 @@
+    @Override
+    public DocumentWorkerDocument validate(final DocumentWorkerDocument document)
+    {
+        final Set<String> keySet = new HashSet<>(document.fields.keySet());
```
Need to make sure that `document.fields` is not null before calling `.keySet()` on it.

jonny-mccoubrey replied:
Behaviour discussed if document.fields is null was to treat the same as if it was an empty list.
How exactly would this be?

dermot-hardy replied:
Currently we're not validating that mandatory fields are present, so this would present as a valid document.  If/when we start to validate the presence of mandatory fields it will move to presenting as invalid at that point.

dermot-hardy commented on worker-batch-ingestion-plugin/src/main/java/com/github/cafdataprocessing/worker/ingestion/validator/FieldValidator.java:
```diff
@@ -0,0 +1,79 @@
+
+        for (final String key : keySet) {
+            try {
+                if (!(isValidField(key))) {
```
Too many brackets:

```suggestion
                if (!isValidField(key)) {
```

dermot-hardy commented on worker-batch-ingestion-plugin/src/main/java/com/github/cafdataprocessing/worker/ingestion/validator/FieldValidator.java:
```diff
@@ -0,0 +1,79 @@
+                    document.fields.remove(key);
+                }
+            } catch (final UnsupportedOperationException ex) {
+                throw new UnsupportedOperationException("Error while modifying immutable Collection", ex);
```
This is not being handled correctly.  You need to substitute the collection with a mutable one.  This should not be a fatal issue.

dermot-hardy commented on worker-batch-ingestion-plugin/src/main/java/com/github/cafdataprocessing/worker/ingestion/validator/FieldValidator.java:
```diff
@@ -0,0 +1,79 @@
+                if (!(isValidField(key))) {
+                    final DocumentWorkerFailure fieldNotAllowedFailure = new DocumentWorkerFailure();
+                    fieldNotAllowedFailure.failureId = "IW-001";
+                    fieldNotAllowedFailure.failureMessage = key + " is not allowed to be set by the agent";
```
Maybe we should also include the value being sent in case it sheds light on this happening unexpectedly.

jonny-mccoubrey replied:
The `key` value is included in the failure message; is this the value you mean?

dermot-hardy replied:
No I was referring to the value.

dermot-hardy commented on worker-batch-ingestion-plugin/src/main/java/com/github/cafdataprocessing/worker/ingestion/validator/FieldValidator.java:
```diff
@@ -0,0 +1,79 @@
+
+        if (document.subdocuments != null) {
+            final ArrayList<DocumentWorkerDocument> subDocs = new ArrayList<>(document.subdocuments);
+            document.subdocuments = new ArrayList<>();
```
What's the purpose of this line, given that you set it after?

dermot-hardy commented on worker-batch-ingestion-plugin/src/main/java/com/github/cafdataprocessing/worker/ingestion/validator/FieldValidator.java:
```diff
@@ -0,0 +1,79 @@
+    public FieldValidator(final String validationFile) throws IOException
+    {
+        final ValidationFileAdapter adapter = new ValidationFileAdapter(validationFile);
+        allowedFields = adapter.getFieldKeys();
```
I'm feeling quite worried about using a list of regular expressions here, because of the performance implications.  We're going to need to evaluate the cost of this approach.  But to give it the best chance possible you should compile the regular expressions ahead of time.

    private final List<Pattern> allowedFieldPatterns;
    ...
    allowedFieldPatterns = adapter.getFieldKeys().stream().map(Pattern::compile).collect(Collectors.toList());

That way the regular expression compilation will only be happening once rather than each time it is used.

Your `isValidField()` method would then become:

    private boolean isValidField(final String fieldKey)
    {
        return allowedFieldPatterns
            .stream()
            .anyMatch(allowedFieldPattern -> allowedFieldPattern.matcher(fieldKey).matches());
    }

dermot-hardy committed:
Code review: Applied a few changes

dermot-hardy committed:
Fixed the test I just broke

dermot-hardy committed:
Fixed the other test I just broke

jonny-mccoubrey committed:
Merge if statements

Co-authored-by: Dermot Hardy <dermot.hardy@microfocus.com>

jonny-mccoubrey committed:
Import regex Pattern lib

jonny-mccoubrey committed:
Add field list size check to testFieldValidator()

jonny-mccoubrey committed:
Code review: Changes

jonny-mccoubrey committed:
Compile regular expressions ahead of time

jonny-mccoubrey committed:
Refactor regex logic

jonny-mccoubrey committed:
Include field value in failure

jonny-mccoubrey commented on worker-batch-ingestion-plugin/src/main/java/com/github/cafdataprocessing/worker/ingestion/validator/FieldValidator.java:
```diff
@@ -0,0 +1,89 @@
+                    fieldNotAllowedFailure.failureMessage = key + " is not allowed to be set by the agent";
+
+                    ObjectMapper mapper = new ObjectMapper();
+                    fieldNotAllowedFailure.failureStack = "Invalid Field Value: "
```
Now adding the value of the field into the failureStack of the Failure incase it helps when unexpected failure occurs. This may not be appropriate place for it, though.

jonny-mccoubrey committed:
Recursively build regex strings

jonny-mccoubrey committed:
Address PR review comments

jonny-mccoubrey committed:
Address PR review comments

dermot-hardy committed:
Merge branch 'develop' into US618160

dermot-hardy committed:
Code review

dermot-hardy committed:
Move ValidationFileAdapter to non-object from interface perspective

dermot-hardy committed:
Code review: Refactoring

dermot-hardy committed:
Code review: Refactoring

dermot-hardy committed:
Code review

dermot-hardy committed:
Code review: Tests
